drivers/SerialBase.cpp

```diff
@@ -120,6 +118,74 @@ int SerialBase::_base_putc(int c)
     return c;
 }
 
+void SerialBase::_init()
+{
+    serial_init(&_serial, _tx_pin, _rx_pin);
+    serial_baud(&_serial, _baud);
+    serial_irq_handler(&_serial, SerialBase::_irq_handler, (uint32_t)this);
+}
+
+void SerialBase::_deinit()
+{
+    serial_free(&_serial);
+}
+
+void SerialBase::_enable_input(bool enable)
+{
+    if (_rx_enabled != enable) {
+        if (enable && !_tx_enabled) {
+            _init();
+        }
+
+        core_util_critical_section_enter();
+        if (enable) {
+            // Enable rx IRQ if attached (indicated by rx IRQ callback not NULL)
+            if (_irq[RxIrq]) {
+                _irq[RxIrq].call();
+                serial_irq_set(&_serial, (SerialIrq)RxIrq, 1);
+            }
+        } else {
+            // Disable rx IRQ
+            serial_irq_set(&_serial, (SerialIrq)RxIrq, 0);
+        }
+        core_util_critical_section_exit();
+
+        _rx_enabled = enable;
+
+        if (!enable && !_tx_enabled) {
+            _deinit();
+        }
+    }
+}
+
+void SerialBase::_enable_output(bool enable)
+{
+    if (_tx_enabled != enable) {
+        if (enable && !_rx_enabled) {
+            _init();
+        }
+
+        core_util_critical_section_enter();
+        if (enable) {
+            // Enable tx IRQ if attached (indicated by tx IRQ callback not NULL)
+            if (_irq[TxIrq]) {
+                _irq[TxIrq].call();
```

Comment:
These automatic IRQ triggers when re-enabled are a good idea, but should be noted in the docs. Just to make sure people aren't too surprised, and also to make it part of the API - guaranteeing an interrupt callback on every enable.

That can simplify code for users if the IRQ call is guaranteed. It simplified `UARTSerial`.